### PR TITLE
Create ScheduledJobIdentityManager interface in IdentityPlugin

### PR DESCRIPTION
**Pull request by cwperks**
### Description

Companion PR in security repo: https://github.com/opensearch-project/security/pull/2773
Companion PR in Job Scheduler: https://github.com/opensearch-project/job-scheduler/pull/394

This PR introduces a new interface inside identity plugin for scheduled job identity management. The new interface is called `ScheduledJobIdentityManager` and contains 3 methods that Job Scheduler will use for extensions use-cases where scheduled jobs run out of process. The current architecture of plugins that extend job scheduler plugin and manage job details in an index owned by the respective plugin (**not** owned by job scheduler, but made aware to job scheduler) can be changed using this new model. 

While the interface is being created here and can be implemented differently, the implementation inside the security plugin being proposed will keep a list of all registered jobs in a single index owned by the security plugin and owner information will be stored in this index. There would be a 1-to-1 mapping of job details stored across all job details indices of plugins that extend job scheduler and this full list owned by the security plugin with owner information. For referential integrity it is stored with the `jobId` and `jobIndex` which is the compound key that links back to the document containing jobs details.

With the user information persisted on a secure index, the implementer of this interface can implement another method `issueAccessTokenOnBehalfOfUser` to issue an access token to be sent to a job runner to be able to execute a job under a user context outside of the user being present.

### Related Issues
Resolves https://github.com/opensearch-project/security/issues/2626

### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed per the DCO using --signoff
- [ ] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/src/main/java/org/opensearch/identity/ScheduledJobIdentityManager.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+ * @opensearch.experimental
+ */
+public interface ScheduledJobIdentityManager {
+
+    /**
+     * Method implemented by an identity plugin to store user information for a scheduled job
+     * @param jobId The id of the scheduled job
+     * @param indexName The index where scheduled job details is stored
+     */
+    void saveUserDetails(String jobId, String indexName);
```
[the remark this answers is not in the document]
Comment (cwperks) [reply server/src/main/java/org/opensearch/identity/ScheduledJobIdentityManager.java:25]: This would be left up to an IdentityPlugin to implement, but the idea with the security plugin is that this user is the creator of the scheduled job. For the security plugin that does mean getting the authenticated user from the threadcontext and persisting it in an index that would contain all scheduled jobs and the user (owner) associated with them.

See https://github.com/opensearch-project/security/pull/2773 for an implementation in the Security plugin.

File: server/src/main/java/org/opensearch/identity/ScheduledJobIdentityManager.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+    /**
+     * Method implemented by an identity plugin to delete user information for a scheduled job
+     * @param jobId The id of the scheduled job
+     * @param indexName The index where scheduled job details is stored
+     */
+    void deleteUserDetails(String jobId, String indexName);
```
[the remark this answers is not in the document]
Comment (cwperks) [reply server/src/main/java/org/opensearch/identity/ScheduledJobIdentityManager.java:32]: This would be if a user deletes a scheduled job altogether. If the original job details is purged from the index that stores the job details then this hook provides a way for an IdentityPlugin to hook into that and react accordingly - for the Security plugin that would mean deleting the entry from the index containing all scheduled jobs and associated user.

**Comment by cwperks** [comment]
@dblock @reta Would love to know your thoughts on this approach of securing scheduled jobs. Please see accompanying Job Scheduler PR for additional details: https://github.com/opensearch-project/job-scheduler/pull/394

**tldr** Centralized scheduled job identity management instead of the current de-centralized management where plugins are responsible for the management and roles injection.

**Review by scrawfor99 — COMMENTED** [review]
Great job Craig! I just left some comments on some style stuff but looks good to me

File: server/src/main/java/org/opensearch/identity/noop/NoopAuthToken.java [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+import org.opensearch.common.io.stream.StreamOutput;
+import org.opensearch.identity.tokens.AuthToken;
+
+import java.io.IOException;
+
+/**
+ * Noop implementation of an Auth Token
+ */
+public class NoopAuthToken extends AuthToken {
```
[the remark this answers is not in the document]
Comment (cwperks) [reply server/src/main/java/org/opensearch/identity/noop/NoopAuthToken.java:19]: It can certainly be switched, Noop in this sense means no auth token which would be the case if an identity plugin was not installed and operating.

File: server/src/main/java/org/opensearch/identity/schedule/ScheduledJobIdentityModel.java [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+import org.opensearch.common.xcontent.XContentParserUtils;
+import org.opensearch.core.xcontent.ToXContentObject;
+import org.opensearch.core.xcontent.XContentBuilder;
+import org.opensearch.core.xcontent.XContentParser;
+
+/**
+ * Class that represents a Scheduled Job Identity
+ */
+public class ScheduledJobIdentityModel implements ToXContentObject {
+    public static final String USER = "user";
```
[the remark this answers is not in the document]
Comment (cwperks) [reply server/src/main/java/org/opensearch/identity/schedule/ScheduledJobIdentityModel.java:22]: IMO these fields should be constants since they define the schema for how this model is structured.

File: server/src/main/java/org/opensearch/identity/schedule/ScheduledJobIdentityModel.java [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+
+    public String getAuthToken() {
+        return this.authToken;
+    }
+
+    public static ScheduledJobIdentityModel parse(final XContentParser parser) throws IOException {
+        ScheduledJobUserModel user = null;
+        String authToken = null;
+
+        XContentParserUtils.ensureExpectedToken(XContentParser.Token.START_OBJECT, parser.currentToken(), parser);
```
[the remark this answers is not in the document]
Comment (cwperks) [reply server/src/main/java/org/opensearch/identity/schedule/ScheduledJobIdentityModel.java:45]: They are all models structured inside of the `operator`. A json-ified version of the model looks like:

```
{
    "operator":
    {
        "user":
        {
            "username": "craig",
            "roles": "all_access",
            "backend_roles": "admin"
        }
    }
}
```

Each nested model should begin with START_OBJECT

File: server/src/main/java/org/opensearch/identity/tokens/BasicAuthToken.java [hunk 6]
```diff
@@ -8,14 +8,18 @@
...
 import java.util.Base64;
 
 /**
  * Basic (Base64 encoded) Authentication Token in a http request header
  */
-public final class BasicAuthToken implements AuthToken {
-
+public final class BasicAuthToken extends AuthToken {
+    public static final String NAME = "basic_auth_token";
     public final static String TOKEN_IDENIFIER = "Basic";
```
[the remark this answers is not in the document]
Comment (cwperks) [reply server/src/main/java/org/opensearch/identity/tokens/BasicAuthToken.java:23]: The NAME here is the name of the NamedWriteable that is serialized for transmission to an extension node.

File: server/src/main/java/org/opensearch/identity/tokens/TokenType.java [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+
+package org.opensearch.identity.tokens;
+
+/**
+ * Enum containing supported token types
+ */
+public enum TokenType {
+    BASIC,
+    BEARER,
+    NONE
```
[the remark this answers is not in the document]
Comment (cwperks) [reply server/src/main/java/org/opensearch/identity/tokens/TokenType.java:17]: None would mean no auth token. If there is no implementation of a ScheduledJobIdentityManager than it would send blank in place of any auth token to the job runner.

**Comment by reta** [comment]
> **tldr** Centralized scheduled job identity management instead of the current de-centralized management where plugins are responsible for the management and roles injection.

@cwperks my apologies if I am missing something, but I do not completely understand how the scheduling should be working here, right now we have a https://github.com/opensearch-project/job-scheduler/ plugin to manage scheduled jobs that it provides a critically important persistence and locking mechanisms (and indeed, this plugin does not support the extensions model where the jobs are out of process).

Now, we have a identity plugin that will be installed on every node on the cluster and will create own scheduler managers that will very likely schedule conflicting executions (since there is no coordination). I am not sure this is the right path to take, looks to me need to focus on general out of process scheduling support first? 

Also, we have this issue which (potentially) make things easier:  https://github.com/opensearch-project/OpenSearch/issues/5310

**Comment by cwperks** [comment]
@reta The IdentityPlugin here is not a replacement for job scheduler, it helps manage the user information that is associated with a scheduled job. Job Scheduler will do all of the job registration, triggering, locking, etc. What the Identity Plugin will do is intercept the job registration process and store the user associated with a job in a centralized index managed by the security plugin.

Right now for anomaly detection, a job details definition looks like:

```
.opendistro-anomaly-detector-jobs
{
    "name": "ad-job-1",
    "schedule":
    {
        "interval":
        {
            "start_time": 16844430461234,
            "period": 10,
            "unit": "Minutes"
        }
    },
    "window_delay":
    {
        "period":
        {
            "interval": 1,
            "unit": "Minutes"
        }
    },
    "enabled": true,
    "enabled_time": 1684443041234,
    "last_update_time": 1684443041234,
    "lock_duration_seconds": 600,
    "user":
    {
        "name": "anomaly_user",
        "backend_roles":
        [
            "backend_role"
        ],
        "roles":
        [
            "all_access"
        ],
        "custom_attribute_names":
        [],
        "user_requested_tenant": "global"
    }
}
```

where user in this document is the user who created the detector. AD with then take advantage of Roles Injection via common-utils to inject these roles into the ThreadContext at job execution. See:

1. Injection in AD: https://github.com/opensearch-project/anomaly-detection/blob/472868d165c2234eece8827584f64086aaf7ae92/src/main/java/org/opensearch/ad/AnomalyDetectorJobRunner.java#L301
2. Roles injection inside sec plugin: https://github.com/opensearch-project/security/blob/main/src/main/java/org/opensearch/security/filter/SecurityFilter.java#L169


What I'm proposing to do here is yank the user out of the job details that AD stores and instead offload that responsibility to the identity plugin. Instead of AD storing the user in its own index with the job details, it would just store the job details.

```
..opendistro-anomaly-detector-jobs
{
    "name": "ad-job-1",
    "schedule":
    {
        "interval":
        {
            "start_time": 16844430461234,
            "period": 10,
            "unit": "Minutes"
        }
    },
    "window_delay":
    {
        "period":
        {
            "interval": 1,
            "unit": "Minutes"
        }
    },
    "enabled": true,
    "enabled_time": 1684443041234,
    "last_update_time": 1684443041234,
    "lock_duration_seconds": 600
}
```

and instead that information would be stored in an index owned by the security plugin:

```
.opendistro-security-scheduled-job-identity
{
  "job_id": "<job_id>",
  "job_index": ".opendistro-anomaly-detector-jobs", // This and the field above are a direct reference to the document in .opendistro-anomaly-detector-jobs index 
  "user":
    {
        "name": "anomaly_user",
        "backend_roles":
        [
            "backend_role"
        ],
        "roles":
        [
            "all_access"
        ],
        "custom_attribute_names":
        [],
        "user_requested_tenant": "global"
    }
}
```

If this is the case then roles injection can be offloaded from each plugin scheduling jobs to the Identity Plugin and for extensions the identity plugin would be the plugin that issues access tokens.

**Comment by reta** [comment]
> @reta The IdentityPlugin here is not a replacement for job scheduler, it helps manage the user information that is associated with a scheduled job. Job Scheduler will do all of the job registration, triggering, locking, etc. What the Identity Plugin will do is intercept the job registration process and store the user associated with a job in a centralized index managed by the security plugin.

Oh I see, thanks @cwperks, since `IdentityService` is now in core, why the plugins couldn't fetch this information? (they have access to the subject) The JobScheduler could use impersonation (or any other suitable techniques) to run the job with the right security context? It seems like this is not just specific to scheduler, fe [notification plugin](https://github.com/opensearch-project/notifications) may need the user context to send the notifications as well.

**Comment by cwperks** [comment]
@reta IMO It shouldn't be the plugin that runs the job's responsibility to inject the roles. Every plugin that wants to add security to their scheduled jobs needs to store the user info alongside the job details in the jobs index that is owned by the plugin. This tightly couples the security plugin and other plugins together. 

A better abstraction is for a central identity system to manage that information. When it comes time to job execution, it should be the identity system's responsibility to declare what user context to run the job with - **not** the plugin that runs the job. For extensions, it certainly shouldn't be up to an extension to declare what roles a job runs with. There would be nothing preventing an extension from saying "run this with all_access" and that should be prohibited. I would not want to carry any analogy of "Role Injection" to extensions. 

Plugins can certainly read Subject information from the threadcontext, but it should be up to the identity system to populate the threadcontext with subject information - not any other plugin.

**Comment by reta** [comment]
> @reta IMO It shouldn't be the plugin that runs the job's responsibility to inject the roles. Every plugin that wants to add security to their scheduled jobs needs to store the user info alongside the job details in the jobs index that is owned by the plugin. This tightly couples the security plugin and other plugins together.

I totally agree with you here, I believe the goal of identity story was/is to bring the identity awareness to the plugins and extensions - in my understanding, it introduces coupling with identity only (security plugin provides identity management as well but using core APIs).

> A better abstraction is for a central identity system to manage that information. When it comes time to job execution, it should be the identity system's responsibility to declare what user context to run the job with - not the plugin that runs the job. For extensions, it certainly shouldn't be up to an extension to declare what roles a job runs with

Totally agree, but how it should be enforced without the plugin/extension participation? For example, let say the user schedules a job passing the security checks (this is job submission identity context). Now, it the `ScheduledJobRunner` (in context of the job-scheduler plugin) could do whatever it wants, right?

**Comment by cwperks** [comment]
> Totally agree, but how it should be enforced without the plugin/extension participation? For example, let say the user schedules a job passing the security checks (this is job submission identity context). Now, it the ScheduledJobRunner (in context of the job-scheduler plugin) could do whatever it wants, right?

That's correct, as of now there is nothing preventing plugins from injecting roles into the threadcontext or stashing their context and assuming super user privileges. Consider this a stepping stone to eventually deprecating the usage of roles injection and threadcontext stashing. With this model of scheduled job security, it gives plugin developers whose plugins schedule jobs an offramp from having to store user information in their jobs index and inject roles at runtime. 

This does mean that there would need to be migrations for each plugin that utilizes roles injection for scheduled job security, but I don't see any other way of securing scheduled jobs for extensions when the user information is stored in an index owned by the plugin/extension. 

I **believe** the migration of existing scheduled jobs should be fairly straightfoward (famous last words). For AD it would mean looping through existing jobs in `.opendistro-anomaly-detector-jobs` and transferring the `user` stored in that index over into a document inside `.opendistro-security-scheduled-job-identity`. I'm envisioning that this system is opt in for plugins where it would need to be explicitly enabled to utilize the identity plugin for management of user's associated with the scheduled jobs.

For extensions this would be the approach to use, the plugin model would be unsupported.

**Comment by reta** [comment]
Thanks @cwperks , but why we are so focused user and roles ? The identity could imply the presence of user or/and roles, but for plugins / extensions I would think about it as a token (which could be encrypted for example), so the plugin / extension (like for example job plugin), could store it and provide the identity when the action (job scheduling, notification, ...) needs to be performed. The identity could be also attached not only to the user but a service and job itself - exactly as you described - but stored as opaque token with the job/notification/etc and verified by identity service (not sure if it makes sense).

**Comment by cwperks** [comment]
@reta that was considered in a previous iteration, but a concern raised was that an extension would be overwrite the opaque token that a job runs with without security in the picture.

Say for example in a design where an extension stores on opaque token with the job details:

1. I login as the `craig` user and create a scheduled job in Anomaly Detection:

```
.opendistro-anomaly-detector-jobs
{
    "name": "ad-job-1",
    "schedule":
    {
        ...
    },
    "token": "<opaque_token_craig>"
}
```

For the sake of this example let's say my user is only allowed to read from `logs-*` index.

2. Admin later comes in and creates a scheduled job:
```
.opendistro-anomaly-detector-jobs
{
    "name": "ad-job-2",
    "schedule":
    {
        ...
    },
    "token": "<opaque_token_admin>"
}
```

What would prevent the extension/plugin from overwriting the token associated with `ad-job-1` with the token that gives admin permissions?

**Comment by reta** [comment]
> What would prevent the extension/plugin from overwriting the token associated with `ad-job-1` with the token that gives admin permissions?

By including the job identity (or in general - any plugin/extension sprcific details) into the token (it could be as simple as just a job name fe or any other techniques like job parameters hash, etc).

**Comment by cwperks** [comment]
What is unique is the combination of `jobId` and `jobIndex`. `jobId` is mostly unique, but it is a documentId corresponding to the document in the plugin's job index and documentIds are not guaranteed to be unique across indices so it needs to be in combination with `jobIndex` to be guaranteed to be unique.

So one idea to prevent a plugin/extension from overwriting an identity associated with a job is to include the `jobId` and `jobIndex` into the tamper-proof token and then use that to compare with the job info. That makes sense to me, but another pro of centralized identity management is being able to develop APIs or UX around managing the identities associated with scheduled jobs. When all of these tokens are stored across disparate indices then it requires knowing about all of them to find the full catalog of tokens that have been issued. IMO I would prefer to centralize anything to do with identity in one place.

**Comment by reta** [comment]
> So one idea to prevent a plugin/extension from overwriting an identity associated with a job is to include the `jobId` and `jobIndex` into the tamper-proof token and then use that to compare with the job info.

👍 

> That makes sense to me, but another pro of centralized identity management is being able to develop APIs or UX around managing the identities associated with scheduled jobs. 

I see pros and cons here. Tokens would need revocations - that the security plugin has to manage centrally, but I am not sure you need the "token" view of the jobs for example. The job-scheduler would have an UI with all the jobs, registered / executed / scheduled / etc, it could also show off the token if necessary (using the `IdentityService` for example to extract the basic information). To your point, the security plugin could keep the centralized place for the identity tokens, but the plugin / extensions build on top of that (by adding  `jobId` and `jobIndex` for job scheduling for example). The revocation of the user identity / service identity token would automatically revoke all derived tokens.

**Comment by cwperks** [comment]
I previously thinking about introducing the concept of a `refresh_token` that would be the opaque token given to a plugin/extension to store along with the jobs details. On job invocation, the `refresh_token` would be traded for a new `refresh_token` and an `access_token` and the `access_token` would be the only that an extension could use to perform rest requests on behalf of the user who the job is configured with.

In this design of centralized management, I am treating the document stored in the `.opendistro-security-scheduled-job-identity` as a refresh token of sorts and what makes it tamper-proof is the fact that its a system index (only super admin can tamper). The information from this comment https://github.com/opensearch-project/OpenSearch/pull/7573#issuecomment-1574045477 would all need to be information stored as claims of a token. 

This has been a great discussion @reta! Its been very helpful in helping to hone in the most future-proof approach to securing scheduled jobs for the long-term. No matter whether the job is running in a plugin (same threadcontext) or externally in an extension.

**Comment by reta** [comment]
> This has been a great discussion @reta! Its been very helpful in helping to hone in the most future-proof approach to securing scheduled jobs for the long-term. No matter whether the job is running in a plugin (same threadcontext) or externally in an extension.

Thanks a lot for patience @cwperks ❤️ , as always

**Review by peternied — COMMENTED** [review]
<HACKER_MODE> What stops a plugin from fetching / updating job information via the new job manager?</>

I think there might be issues with dangling references from jobs that failed to delete, or credentials that were updated after a job was recycled.  How do you think we can shore up for these scenarios?

File: server/src/main/java/org/opensearch/identity/schedule/ScheduledJobIdentityModel.java [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+public class ScheduledJobIdentityModel implements ToXContentObject {
+    public static final String USER = "user";
+
+    public static final String TOKEN = "token";
+    private final ScheduledJobUserModel user;
+    private final String authToken;
+
+    public ScheduledJobIdentityModel(final ScheduledJobUserModel user, final String authToken) {
+        this.user = user;
+        this.authToken = authToken;
```
[the remark this answers is not in the document]
Comment (cwperks) [reply server/src/main/java/org/opensearch/identity/schedule/ScheduledJobIdentityModel.java:30]: In the case of plugins, they could pass the `user` in here as its currently stored in many plugins jobs indices. For extensions, they would pass the token since they do not have access to user info on the threadcontext but the token represents the same info.

At REST this data is stored in a system index owned by the security plugin which means that no user, even admin can tamper. It does leave the door open for super admin to modify the identity associated with the job which IMO is desirable. That would allow orphaned references to deleted users or roles to be updated in a central place where its currently confusing being stored in different plugin indices in possibly different ways.

File: server/src/main/java/org/opensearch/identity/tokens/AuthToken.java [hunk 9]
```diff
@@ -8,11 +8,21 @@
...
 
+import org.opensearch.common.io.stream.NamedWriteable;
+
 /**
  * Interface for all token formats to support to authenticate user such as UserName/Password tokens, Access tokens, and more.
  *
  * @opensearch.experimental
  */
-public interface AuthToken {
+public abstract class AuthToken implements NamedWriteable {
```
[the remark this answers is not in the document]
Comment (cwperks) [reply server/src/main/java/org/opensearch/identity/tokens/AuthToken.java:18]: Will do

**Comment by cwperks** [comment]
> <HACKER_MODE> What stops a plugin from fetching / updating job information via the new job manager?</>

Plugins already have the capability to stash their thread context and do anything in the current plugin security model of OpenSearch - plugins have some level of implicit trust that they will do the right thing. IMO we need to start creating new security models that lead to more secure plugins and this is one of those stepping stones that will almost eliminate the need to do roles injection anywhere in plugins.

In addition to that, this manager provides no way of getting the identity stored with the job. It has a function that populates the current execution context with identity (`issueAccessTokenOnBehalfOfUser`) that could inject roles into the thread context (for jobs running in plugins) or issue and send an access token (for jobs running in extensions)

> I think there might be issues with dangling references from jobs that failed to delete, or credentials that were updated after a job was recycled. How do you think we can shore up for these scenarios?

I think its possible to provide some level of atomicity - I will look more into that today with action listeners.

File: server/src/main/java/org/opensearch/identity/tokens/AuthToken.java [hunk 10]
```diff
@@ -8,11 +8,21 @@
...
+
+    public TokenType getTokenType() {
+        return TokenType.NONE;
+    }
+
+    public String getTokenValue() {
```
[the remark this answers is not in the document]
Comment (peternied) [reply server/src/main/java/org/opensearch/identity/tokens/AuthToken.java:24]: > what are the constraints on how this should work

I was thinking around the lines of should this value be cryptographically protected, or can it be in plain text?

There are sometimes when we need to be access the data in memory and we need a handle of the string that was presented that includes the password, but converting from in memory to something we serialize seems like it should have clear usage associated with it such as expiration, encryption.

I think it makes sense to turn a token into something that is serialized, but I think we might want to be very careful when this is done and it might be worthwhile to even wrap these tokens in another secure envelope to ensure they aren't tampered or readable outside the expected scopes.

File: server/src/main/java/org/opensearch/identity/schedule/ScheduledJobIdentityModel.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (peternied) [reply server/src/main/java/org/opensearch/identity/schedule/ScheduledJobIdentityModel.java:30]: Lets go to the high level scenario for a moment because I think it will better inform what data we will have available.

1) As an extension, after I've been initialized, create a job that run as a service account.
2) As an extension, after I've been called by a user, create a job that runs as that user
3) As an extension, after I've been called by a user, create a job that runs as a user other than the user making the call.

For the API call somewhere that creates a job, it takes its existing parameters and it also takes:
- Nothing, and the subject of the current call is used (Covers 1 & 2)
- Principal associated with the job

For the current caller, all the AuthZ information has been loaded, so that works with exactly what you've got planned.

For a passed in principal (Alice wants the job to run as Bob), I think we need a way to generate/hydrate the AuthZ information from a principal - of course checking if this is allowed.

---

However! There is a problem even if we are reusing the current caller, if the AuthZ information expires how do we get it back - which also overlaps with the passed in principal problem space.

Setting aside the details of what is stored within the thread context or how its used.  How can we go from expired authz information -> fresh authz information without the user logging back in?

[the remark this answers is not in the document]
Comment (cwperks) [reply server/src/main/java/org/opensearch/identity/schedule/ScheduledJobIdentityModel.java:30]: > Setting aside the details of what is stored within the thread context or how its used. How can we go from expired authz information -> fresh authz information without the user logging back in?

In this case its driven by the entry in the centralized scheduled job identity index. If a job is scheduled successfully, an entry is added to the secure and centralized `.opendistro-security-scheduled-job-identity` index. You can think of that entry like a refresh token and its (mostly) tamper proof because its in a system index.

```
.opendistro-security-scheduled-job-identity
{
  "job_id": "<job_id>",
  "job_index": ".opendistro-anomaly-detector-jobs", // This and the field above are a direct reference to the document in .opendistro-anomaly-detector-jobs index 
  "user":
    {
        "name": "anomaly_user",
        "backend_roles":
        [
            "backend_role"
        ],
        "roles":
        [
            "all_access"
        ],
        "custom_attribute_names":
        [],
        "user_requested_tenant": "global"
    }
}
```

This entry drives the creation of fresh authz information on job invocation. For plugins that means injecting roles into the threadcontext, for extensions that means issuing a token with the issue where the entry in the `.opendistro-security-scheduled-job-identity` contains the info needed for token issuance like the username (principal) and authorization info like the mapped roles.

[the remark this answers is not in the document]
Comment (peternied) [reply server/src/main/java/org/opensearch/identity/schedule/ScheduledJobIdentityModel.java:30]: The `issueAccessTokenOnBehalfOfUser` method uses the refresh token by reading from `.opendistro-security-scheduled-job-identity` to the authenticating backend to get fresh authz information.  The example above shows the authz information in the new index, is that the refresh token?  How is that used to refresh?

[the remark this answers is not in the document]
Comment (cwperks) [reply server/src/main/java/org/opensearch/identity/schedule/ScheduledJobIdentityModel.java:30]: Yes treat that entry as a refresh token of sorts - its not a JWT but by having the entry there its assumed that a user has previously given consent to have a job run in the background on their behalf.

Its designed this way to have a migration path from how plugins are scheduling jobs today and makes it pretty clear how to convert a plugin with scheduled jobs (like AD) to an extension. 

Plugins will already store the info above alongside the job information in their jobs index and then inject those `roles` into the threadcontext. By migrating that info from the plugin's job index into a centralized system index that is tamper proof - it is more secure and offloads the work of roles injection from the plugin/extension to the identity system.

It may not be encoded and guaranteed to be tamper-proof like a JWT, but since it is a system index there are guarantees that it cannot be meddled with by anyone with very few exceptions. By storing it like this, it would make it easier to develop APIs are changing the identity or roles that a job runs with.

[the remark this answers is not in the document]
Comment (peternied) [reply server/src/main/java/org/opensearch/identity/schedule/ScheduledJobIdentityModel.java:30]: A refresh token is used to create fresh AuthZ information, I don't see how this is possible from the raw information that is placed in the index except for basic auth users.  Please update this thread with data flow diagrams or links to the code that illustrates how this will work.

[the remark this answers is not in the document]
Comment (cwperks) [reply server/src/main/java/org/opensearch/identity/schedule/ScheduledJobIdentityModel.java:30]: Not sure I am following. The entry in the scheduled job identity index contains all of the information to create new access tokens. If we were to store a refresh token, all of the same info would need to be stored as claims in the refresh token.

```
.opendistro-security-scheduled-job-identity
{
  "job_id": "<job_id>",
  "job_index": ".opendistro-anomaly-detector-jobs", // This and the field above are a direct reference to the document in .opendistro-anomaly-detector-jobs index 
  "user":
    {
        "name": "anomaly_user",
        "backend_roles":
        [
            "backend_role"
        ],
        "roles":
        [
            "all_access"
        ],
        "custom_attribute_names":
        [],
        "user_requested_tenant": "global"
    }
}
```

When creating a new access token, the username (subject) and mapped roles are required to make this backward compatible with how plugins do security for scheduled jobs.

**Comment by cwperks** [comment]
Closing for now. Will re-open at a later date with updates.
